### PR TITLE
misc: timings script

### PR DESCRIPTION
**Pull request by connorjclark**
For example usages, see #9712 and #9713

### CODE REVIEW

File: lighthouse-core/scripts/timings.js [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    'summarize': 'Prints statistics report',
+    'measure-filter': 'Regex filter of measures to report. Optional',
+    'output': 'table, json',
+  })
+  .string('measure-filter')
+  .default('output', 'table')
+  .array('urls')
+  .string('lh-flags')
+  .default('lh-flags', '')
+  // Why is the printing for examples so awful?
```
Comment (connorjclark) [inline lighthouse-core/scripts/timings.js]: Anyone know how to make yargs examples output not look terrible?
image

Comment (patrickhulce) [reply lighthouse-core/scripts/timings.js]: We're still on *ancient* yargs so I'm sure it's been fixed in a version released in the past 5 years :)

File: lighthouse-core/scripts/timings.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  // eslint-disable max-len
+  // .example('node lighthouse-core/scripts/timings.js --name my-collection --collect -n 3 --lh-flags=\'--only-audits=unminified-javascript\' --urls https://www.example.com', 'Collect')
+  // .example('node lighthouse-core/scripts/timings.js --name my-collection --summarize --measure-filter \'loadPage|connect\'', 'Summarize')
+  // eslint-enable max-len
+  .wrap(yargs.terminalWidth())
+  .argv;
+
+const outputDir = `${ROOT_OUTPUT_DIR}/${argv.name}`;
+
+if (argv.collect) {
```
Comment (patrickhulce) [inline lighthouse-core/scripts/timings.js]: might be nice to have a `main()`, `collect()` and `summarize()` to separate this core logic from the helpers :)

File: lighthouse-core/scripts/timings.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+      measure: measureName,
+      url,
+      n: measures.length,
+      mean: round(mean),
+      stdev: round(stdev),
+      min,
+      max,
+    };
+  }).sort((a, b) => {
+    return a.measure.localeCompare(b.measure);
```
Comment (patrickhulce) [inline lighthouse-core/scripts/timings.js]: secondary sort on `url` if that's the other unique piece?

File: lighthouse-core/scripts/timings.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    const min = Math.min(...measures);
+    const max = Math.max(...measures);
+    const stdev = Math.sqrt(average(measures.map(measure => (measure - mean) ** 2)));
+    return {
+      measure: measureName,
+      url,
+      n: measures.length,
+      mean: round(mean),
+      stdev: round(stdev),
+      min,
```
Comment (patrickhulce) [inline lighthouse-core/scripts/timings.js]: round these bad boys too? :)

Comment (connorjclark) [reply lighthouse-core/scripts/timings.js]: they actually already get truncated to hundredths (https://github.com/GoogleChrome/lighthouse/blob/266404d2b2d85c2e5ad7a770a12ab0adf7cf7830/lighthouse-core/runner.js#L190) so I didn't notice, but yeah for consistency will round these too

File: lighthouse-core/scripts/timings.js [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+      measures.push(measureEntry.duration);
+    }
+  }
+
+  const results = [...measuresMap.entries()].map(([measuresKey, measures]) => {
+    const [url, measureName] = measuresKey.split('@@@');
+    const mean = average(measures);
+    const min = Math.min(...measures);
+    const max = Math.max(...measures);
+    const stdev = Math.sqrt(average(measures.map(measure => (measure - mean) ** 2)));
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/scripts/timings.js]: for those wondering why:

https://www.khanacademy.org/math/statistics-probability/summarizing-quantitative-data/variance-standard-deviation-sample/a/population-and-sample-standard-deviation-review

https://www.khanacademy.org/math/ap-statistics/summarizing-quantitative-data-ap/more-standard-deviation/v/review-and-intuition-why-we-divide-by-n-1-for-the-unbiased-sample-variance?qa_expand_key=kaencrypted_47b01a4d47efe629deb370110a0ea31c_ed01653c7d4a99f89435928e0d0bb81f86a2d612c1ec4114d49fab4b4a8ae325873f466516417a74ec6b9f8739c3e4a481e1564411db72e9b77c7e9463c2d09846a98e9a62d5951a60d24f64329cadf9c45cbcbce32f4fad9bece943cdf43db6d6f9452d3af59b1c34fb848d5d02e79e750c4ed7daf3d483852c11dda1f71de579251a756354f1f528a1f04746257c3277a22ebeaba8352107bcd77f7cf0a94593bd607fc615be68321168cbc0f575fe60bb3cbf6c78f9bbeeb870f3ce48eae7

File: lighthouse-core/scripts/timings.js [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+      }
+
+      const measuresKey = `${lhr.requestedUrl}@@@${measureName}`;
+      let measures = measuresMap.get(measuresKey);
+      if (!measures) {
+        measures = [];
+        measuresMap.set(measuresKey, measures);
+      }
+
+      const measureEntry = lhr.timing.entries.find(measure => measure.name === measureName);
```
Comment (connorjclark) [inline lighthouse-core/scripts/timings.js]: wow what the heck am I doing here, this only ever grabs the first of each measure name.

File: lighthouse-core/scripts/timings.js [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+      if (!measureEntry) throw new Error('missing measure');
+
+      measures.push(measureEntry.duration);
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/scripts/timings.js]: cool, done

File: lighthouse-core/scripts/timings.js [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+  const measuresMap = new Map();
+  /** @type {RegExp|null} */
+  const measureFilter = argv.measureFilter ? new RegExp(argv.measureFilter, 'i') : null;
+
+  for (const lhrPath of fs.readdirSync(outputDir)) {
+    const lhrJson = fs.readFileSync(`${outputDir}/${lhrPath}`, 'utf-8');
+    /** @type {LH.Result} */
+    const lhr = JSON.parse(lhrJson);
+
+    for (const measureName of lhr.timing.entries.map(entry => entry.name)) {
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/scripts/timings.js]: gave it a shot

**Comment by connorjclark** [comment]
`compare-timings.js`?

**Comment by brendankenny** [comment]
> compare-timings.js?

SGTM

File: lighthouse-core/scripts/compare-timings.js [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+  /** @type {RegExp|null} */
+  const measureFilter = argv.measureFilter ? new RegExp(argv.measureFilter, 'i') : null;
+
+  for (const lhrPath of fs.readdirSync(outputDir)) {
+    const lhrJson = fs.readFileSync(`${outputDir}/${lhrPath}`, 'utf-8');
+    /** @type {LH.Result} */
+    const lhr = JSON.parse(lhrJson);
+
+    // Sum all measures of the same name.
+    const measuresSummed = lhr.timing.entries.reduce((acc, entry) => {
```
Comment (brendankenny) [inline lighthouse-core/scripts/compare-timings.js]: nit: a regular for loop is probably a better fit here

File: lighthouse-core/scripts/compare-timings.js [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+      ].join(' ');
+      execSync(cmd, {stdio: 'ignore'});
+    }
+  }
+}
+
+function summarize() {
+  // `${url}@@@${entry.name}` -> duration
+  /** @type {Map<string, number[]>} */
+  const durationsMap = new Map();
```
Comment (brendankenny) [inline lighthouse-core/scripts/compare-timings.js]: not sure this is better. Now we have `timing`, `measure`, and `duration` :)

File: lighthouse-core/scripts/compare-timings.js [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+  /** @type {RegExp|null} */
+  const measureFilter = argv.measureFilter ? new RegExp(argv.measureFilter, 'i') : null;
+
+  for (const lhrPath of fs.readdirSync(outputDir)) {
+    const lhrJson = fs.readFileSync(`${outputDir}/${lhrPath}`, 'utf-8');
+    /** @type {LH.Result} */
+    const lhr = JSON.parse(lhrJson);
+
+    // Sum all measures of the same name.
```
Comment (brendankenny) [inline lighthouse-core/scripts/compare-timings.js]: this variable has no summing :)

File: lighthouse-core/scripts/compare-timings.js [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+    // Push the average of all the measures of each name.
+    for (const [name, durationsForSingleRun] of Object.entries(measuresSummed)) {
+      const measuresKey = `${lhr.requestedUrl}@@@${name}`;
+      let durations = durationsMap.get(measuresKey);
+      if (!durations) {
+        durations = [];
+        durationsMap.set(measuresKey, durations);
+      }
+      durations.push(average(durationsForSingleRun));
```
Comment (brendankenny) [inline lighthouse-core/scripts/compare-timings.js]:
```suggestion
      durations.push(sum(durationsForSingleRun));
```
I assume?

Comment (connorjclark) [reply lighthouse-core/scripts/compare-timings.js]: yea thats right

**Pull request merged**